### PR TITLE
Additions to symengine/sympy_compat.py

### PR DESCRIPTION
**Pull request by ShikharJ**
Relevant: #136 
@isuruf The constructor for `Float` is itself quite complex and many of the functionalities used are currently unavailable. I have commented the specific parts as of now. What should be done here?

### CODE REVIEW

**Comment by ShikharJ** [comment]
Ping @isuruf.

File: symengine/sympy_compat.py [hunk 1]
```diff
@@ -43,6 +44,143 @@ def __new__(cls, i):
         return symengine.Integer(i)
 
 
+class Float(Number):
+    _classes = (symengine.RealDouble,)
+
+    def __new__(cls, num, dps=None, prec=None, precision=None):
+        if prec is not None:
```
Comment (isuruf) [inline symengine/sympy_compat.py]: You can remove this. It's already deprecated in sympy

**Comment by isuruf** [comment]
What are you planning to do next?

**Comment by ShikharJ** [comment]
@isuruf My short term goal for the next 4-5 days would be to get this and #129 merged in. After that, I'd like to devote my attention towards porting `SymEngine` in `sympy/parsing` and `sympy/physics` modules (I'll continue on with this till the official coding period begins).

**Comment by isuruf** [comment]
Thanks for the update. I particularly meant what are you going to do next in this PR? Float class doesn't really do anything here

**Comment by ShikharJ** [comment]
That would be one of my doubts. We're currently using `RealMPFR` for `sympy.Float` objects. Should I try interfacing `RealMPFR` in place of `mpmath` objects? Would that be a viable plan?

**Comment by isuruf** [comment]
Yes, Float class should return a `RealDouble` if the precision is <= 53 bits and a `RealMPFR` if not.

File: symengine/sympy_compat.py [hunk 2]
```diff
@@ -43,6 +67,137 @@ def __new__(cls, i):
...
+                except decimal.InvalidOperation:
+                    pass
+                else:
+                    isint = '.' not in num
+                    num, dps = _decimal_to_Rational_prec(Num)
+                    if num.is_Integer and isint:
+                        dps = max(dps, len(str(num).lstrip('-')))    
+                    dps = max(15, dps)
+                    '''
+                    precision = mlib.libmpf.dps_to_prec(dps)
```
Comment (ShikharJ) [inline symengine/sympy_compat.py]: @isuruf Would you be fine with re-implementing `dps_to_prec` in `symengine.py`? Or would it cause a problem due to the licenses?

Comment (isuruf) [inline symengine/sympy_compat.py]: I think `dps_to_prec` is a one liner and that's fine

File: symengine/sympy_compat.py [hunk 3]
```diff
@@ -43,6 +67,137 @@ def __new__(cls, i):
...
+            num = num._mpf_
+        '''
+        
+        if dps is None and precision is None:
+            dps = 15
+            if isinstance(num, Float):
+                return num
+            if isinstance(num, str) and _literal_float(num):
+                try:
+                    Num = decimal.Decimal(num)
```
Comment (isuruf) [inline symengine/sympy_compat.py]: We don't need `Decimal` support.

**Comment by isuruf** [comment]
I think you are trying too much to do what sympy does.
Do something like this,
1. convert dps to prec
2. If prec <= 53 return RealDouble(str(num))
3. Else return RealMPFR(str(num), prec))

File: symengine/sympy_compat.py [hunk 4]
```diff
@@ -43,6 +47,35 @@ def __new__(cls, i):
...
+    _classes = (symengine.RealDouble, symengine.RealMPFR)
+
+    def __new__(cls, num, dps=None, prec=None, precision=None):
+        if dps is not None and precision is not None:
+            raise ValueError('Both decimal and binary precision supplied. '
+                             'Supply only one. ')
+        if dps is None and precision is None:
+            dps = 15
+            if isinstance(num, Float):
+                return num
```
Comment (isuruf) [inline symengine/sympy_compat.py]: This is wrong, what if `num` had a different precision?

File: symengine/sympy_compat.py [hunk 5]
```diff
@@ -43,6 +47,35 @@ def __new__(cls, i):
         return symengine.Integer(i)
 
 
+class Float(Number):
+    _classes = (symengine.RealDouble, symengine.RealMPFR)
+
+    def __new__(cls, num, dps=None, prec=None, precision=None):
```
Comment (isuruf) [inline symengine/sympy_compat.py]: `prec` is not used. You can remove that

File: symengine/sympy_compat.py [hunk 6]
```diff
@@ -43,6 +47,35 @@ def __new__(cls, i):
...
+            raise ValueError('Both decimal and binary precision supplied. '
+                             'Supply only one. ')
+        if dps is None and precision is None:
+            dps = 15
+            if isinstance(num, Float):
+                return num
+        elif precision == '' and dps is None or precision is None and dps == '':
+            if not isinstance(num, (str, unicode)):   
+                raise ValueError('The null string can only be used when '
+                'the number to Float is passed as a string or an integer.')
```
Comment (isuruf) [inline symengine/sympy_compat.py]: I don't get what the purpose here is.

File: symengine/sympy_compat.py [hunk 7]
```diff
@@ -43,6 +47,35 @@ def __new__(cls, i):
...
+        if precision is None or precision == '':
```
Comment (isuruf) [inline symengine/sympy_compat.py]: Remove `precision == ''` part. It's unnecessary

File: symengine/sympy_compat.py [hunk 8]
```diff
@@ -43,6 +47,35 @@ def __new__(cls, i):
...
+            precision = dps_to_prec(dps)
+        if HAVE_SYMENGINE_MPFR:
+            if precision > 53:
+                return RealMPFR(str(num), precision)    
+            else:
+                return RealDouble(float(str(num)))
+        else:
+            RealDouble(float(str(num)))
```
Comment (isuruf) [inline symengine/sympy_compat.py]: This should throw an error if `precision > 53`

File: symengine/sympy_compat.py [hunk 9]
```diff
@@ -43,6 +47,35 @@ def __new__(cls, i):
...
+            if not isinstance(num, (str, unicode)):   
+                raise ValueError('The null string can only be used when '
+                'the number to Float is passed as a string or an integer.')
+        if precision is None or precision == '':
+            precision = dps_to_prec(dps)
+        if HAVE_SYMENGINE_MPFR:
+            if precision > 53:
+                return RealMPFR(str(num), precision)    
+            else:
+                return RealDouble(float(str(num)))
```
Comment (isuruf) [inline symengine/sympy_compat.py]: This can be `RealDouble(float(num))`, sorry about that

File: symengine/sympy_compat.py (hunk 4 above)
Comment (ShikharJ) [inline symengine/sympy_compat.py]: Line 57 takes that into account. Rest cases have been accounted for later.

File: symengine/sympy_compat.py [hunk 10]
```diff
@@ -43,6 +47,53 @@ def __new__(cls, i):
...
+                    return RealDouble(num.__float__())
+                else:
+                    return RealDouble(float(num))
+        else:
+            if precision > 53:
+                raise ValueError('RealMPFR unavailable for high precision numerical values.')
+            else:
+                if isinstance(num, RealDouble):
+                    return RealDouble(num.__float__())
+                elif isinstance(num, RealMPFR):
```
Comment (isuruf) [inline symengine/sympy_compat.py]: This case is impossible

File: symengine/sympy_compat.py [hunk 11]
```diff
@@ -43,6 +47,53 @@ def __new__(cls, i):
...
+
+        if HAVE_SYMENGINE_MPFR:
+            if precision > 53:
+                if isinstance(num, RealDouble):
+                    return RealMPFR(str(num.__float__()), precision)
+                elif isinstance(num, RealMPFR):
+                    if precision == num.get_prec():
+                        return num
+                    else:
+                        return RealMPFR(str(num.__float__()), precision)
```
Comment (isuruf) [inline symengine/sympy_compat.py]: `num.__float__()` loses precision

File: symengine/sympy_compat.py [hunk 12]
```diff
@@ -43,6 +47,53 @@ def __new__(cls, i):
...
+            dps = 15
+            if isinstance(num, Float):
+                return num
+        if precision is None:
+            precision = dps_to_prec(dps)
+
+        if HAVE_SYMENGINE_MPFR:
+            if precision > 53:
+                if isinstance(num, RealDouble):
+                    return RealMPFR(str(num.__float__()), precision)
```
Comment (isuruf) [inline symengine/sympy_compat.py]: Remove `__float__()` here

File: symengine/sympy_compat.py [hunk 13]
```diff
@@ -43,6 +47,53 @@ def __new__(cls, i):
...
+                        return num
+                    else:
+                        return RealMPFR(str(num.__float__()), precision)
+                else:
+                    return RealMPFR(str(num), precision)    
+            else:
+                if isinstance(num, RealDouble):
+                    return num
+                elif isinstance(num, RealMPFR):
+                    return RealDouble(num.__float__())
```
Comment (isuruf) [inline symengine/sympy_compat.py]: Use `float(num)` instead of `num.__float__()`

File: symengine/sympy_compat.py [hunk 14]
```diff
@@ -43,6 +47,53 @@ def __new__(cls, i):
...
+                else:
+                    return RealDouble(float(num))
+        else:
+            if precision > 53:
+                raise ValueError('RealMPFR unavailable for high precision numerical values.')
+            else:
+                if isinstance(num, RealDouble):
+                    return RealDouble(num.__float__())
```
Comment (isuruf) [inline symengine/sympy_compat.py]: Just return `num` here.

File: symengine/sympy_compat.py (hunk 4 above)
Comment (isuruf) [inline symengine/sympy_compat.py]: No, it doesn't. `dps = 15` is not honoured here.

Comment (ShikharJ) [inline symengine/sympy_compat.py]: `SymPy` sets `dps = 15` for cases where the argument is not an instance of `Float`. If the `num` is an instance of the `Float` class and no `dps` or `precision` is provided, it simply returns that same object. Should I change this? `dps = 15` evaluates to `precision = 53`, so should I return a RealDouble each time?

Comment (isuruf) [inline symengine/sympy_compat.py]: Yes, that would be consistent. Also, you can just remove the 2 lines above and the rest of the code below should handle that case.

**Comment by ShikharJ** [comment]
@isuruf The builds are failing because the `HAVE_SYMENGINE_FLAG` is not available to `sympy_compat.py` and the `test_sympy_compat.py`. What can be done here?

**Comment by isuruf** [comment]
@ShikharJ, use `have_mpfr` in tests

**Comment by isuruf** [comment]
Thanks for the PR

**Comment by ShikharJ** [comment]
Thanks for the extensive reviewing @isuruf. You're the best!

**Pull request merged**